File: 42_trapping_rain_water/solution.ts
/**
 * Time: O(n) | Space: O(n)
 */
function trap1(height: number[]): number {
  const maxL = Array.from({ length: height.length }, () => 0);
  const maxR = Array.from({ length: height.length }, () => 0);

  for (let i = 1; i < height.length; i++) {
    if (height[i - 1] > maxL[i - 1]) {
      maxL[i] = height[i - 1];
    } else {
      maxL[i] = maxL[i - 1];
    }
  }

  for (let i = height.length - 2; i >= 0; i--) {
    if (height[i + 1] > maxR[i + 1]) {
      maxR[i] = height[i + 1];
    } else {
      maxR[i] = maxR[i + 1];
    }
  }

  let sum = 0;

  for (let i = 0; i < height.length; i++) {
    const water = Math.max(Math.min(maxL[i], maxR[i]) - height[i], 0);
    sum += water;
  }

  return sum;
}

/**
 * Time: O(n) | Space: O(1)
 */
function trap2(height: number[]): number {
  let sum = 0;
  let maxL = height[0];
  let maxR = height[height.length - 1];
  let left = 0;
  let right = height.length - 1;

  while (left < right) {
    if (maxL <= maxR) {
      left++;

      const water = Math.max(Math.min(maxL, maxR) - height[left], 0);
      sum += water;

      if (height[left] > maxL) {
        maxL = height[left];
      }
    } else {
      right--;

      const water = Math.max(Math.min(maxL, maxR) - height[right], 0);
      sum += water;

      if (height[right] > maxR) {
        maxR = height[right];
      }
    }
  }

  return sum;
}

File: 42_trapping_rain_water/solution.py
from typing import List


class Solution:
    # Time: O(n), Space: O(n)
    def trap(self, height: List[int]) -> int:
        maxL, maxR = [0] * len(height), [0] * len(height)
        water = [0] * len(height)

        # max left
        for i in range(1, len(height)):
            maxL[i] = max(maxL[i - 1], height[i - 1])

        # max right
        for i in range(len(height) - 2, -1, -1):
            maxR[i] = max(maxR[i + 1], height[i + 1])

        # # calculate the water in each position
        for i in range(len(height)):
            w = min(maxL[i], maxR[i]) - height[i]
            if w > 0:
                water[i] = w

        return sum(water)

    # Time: O(n), Space: O(1)
    def trap2(self, height: List[int]) -> int:
        water = 0
        l, r = 0, len(height) - 1
        maxL, maxR = height[l], height[r]

        while l < r:
            if maxL <= maxR:
                l += 1
                sum = maxL - height[l]
                if sum > 0:
                    water += sum
                maxL = max(maxL, height[l])

            else:
                r -= 1
                sum = maxR - height[r]
                if sum > 0:
                    water += sum
                maxR = max(maxR, height[r])

        return water
